### PR TITLE
[Git] Upgrade to v2.36.1 and add hack for macOS

### PR DESCRIPTION
**Pull request by giordano**
CC: @mortenpi 

Ref https://github.com/JuliaVersionControl/Git.jl/issues/40.

### CODE REVIEW

**Comment by staticfloat** [comment]
Does this not work properly?

**Comment by giordano** [comment]
The build per se works fine, but I also wanted to address https://github.com/JuliaVersionControl/Git.jl/issues/40. However last time I tried it, it didn't seem to work around the SIP limitation, there's a build at https://github.com/giordano/Git_jll.jl if you want to play with it.

File: G/Git/build_tarballs.jl [hunk 1]
```diff
@@ -43,24 +44,49 @@ else
...
+    for file in $(grep -l -r '^#!/bin/sh' "${prefix}"); do
+        if [[ $(basename "${file}") == git* ]]; then
+            sed -i -e 's?^#!/bin/sh?#!/bin/sh\nexport DYLD_FALLBACK_LIBRARY_PATH="${JLL_DYLD_FALLBACK_LIBRARY_PATH}"?' "${file}"
+        fi
+    done
+    for file in $(grep -l -r '^#!/usr/bin/perl' "${prefix}"); do
+        if [[ $(basename "${file}") == git* ]]; then
+            sed -i -e 's?^#!/usr/bin/perl?#!/usr/bin/perl\n$ENV{'DYLD_FALLBACK_LIBRARY_PATH'} = $ENV{'JLL_DYLD_FALLBACK_LIBRARY_PATH'};?' "${file}"
+        fi
+    done
```
Comment (staticfloat) [inline G/Git/build_tarballs.jl:74]: I think with the `git` renaming, you don't need these lines anymore

Comment (giordano) [reply G/Git/build_tarballs.jl]: Ok, yes, I can confirm it works also without these lines.

**Comment by giordano** [comment]
Uhm, this isn't working though:
```
julia> run(`$(git()) credential-osxkeychain`)
dyld[2558]: Library not loaded: '@rpath/libpcre2-8.0.dylib'
  Referenced from: '/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/bin/_git'
  Reason: tried: '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/bin/../lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/bin/../lib/libpcre2-8.0.dylib' (no such file), '/libpcre2-8.0.dylib' (no such file)
ERROR: failed process: Process(setenv(`/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/bin/git credential-osxkeychain`,["XPC_FLAGS=0x0", "PATH=/Users/mose/.julia/artifacts/41d2d4f312d93b56961767987cf48ea1f6d6fadc/bin:/Users/mose/.julia/artifacts/b1a055a9c307b137556560efcc2cee3a6405dfd7/bin:/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/bin:/Users/mose/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/mose/.cargo/bin", "COMMAND_MODE=unix2003", "PWD=/Users/mose", "XPC_SERVICE_NAME=0", "TERM_PROGRAM=iTerm.app", "HISTSIZE=1000000", "GIT_EXEC_PATH=/Users/mose/.julia/artifacts/1e557aed8ecb7576d4215c8ef8670079e81c11ef/libexec/git-core", "SHELL=/bin/zsh", "DYLD_FALLBACK_LIBRARY_PATH=/Users/mose/repo/julia/usr/lib:/Users/mose/.julia/artifacts/41d2d4f312d93b56961767987cf48ea1f6d6fadc/lib:/Users/mose/.julia/artifacts/b1a055a9c307b137556560efcc2cee3a6405dfd7/lib:/Users/mose/.julia/artifacts/b376941014acf8e5501996cdf9932036cfb3bb71/lib:/Users/mose/repo/julia/usr/bin/../lib/julia:/Users/mose/repo/julia/usr/bin/../lib:/Users/mose/lib:/usr/local/lib:/lib:/usr/lib"  …  "USER=mose", "JULIA_PKG_USE_CLI_GIT=true", "ITERM_SESSION_ID=w0t0p0:C27C9DE9-878A-4627-8A7B-9C9C7CE07D2F", "GIT_SSL_CAINFO=/Users/mose/repo/julia/usr/share/julia/cert.pem", "TERM=xterm-256color", "HOME=/Users/mose", "TERM_PROGRAM_VERSION=3.4.19", "JULIA_PKG_SERVER_REGISTRY_PREFERENCE=eager", "COLORTERM=truecolor", "OPENBLAS_MAIN_FREE=1"]), ProcessSignaled(6)) [0]

Stacktrace:
 [1] pipeline_error
   @ ./process.jl:565 [inlined]
 [2] run(::Cmd; wait::Bool)
   @ Base ./process.jl:480
 [3] run(::Cmd)
   @ Base ./process.jl:477
 [4] top-level scope
   @ REPL[5]:1
```

**Comment by giordano** [comment]
Ah, nevermind, I was using vanilla `Git.jl`, which of course at the moment doesn't set `JLL_DYLD_FALLBACK_LIBRARY_PATH`.  If I manually do that, then it works.

**Comment by staticfloat** [comment]
Ah, I was juuuuust about to post that! :)

**Comment by giordano** [comment]
Good to go now? :slightly_smiling_face:

**Review by staticfloat — APPROVED** [review]
SGTM, I'll let you merge when you think enough other people have tried it out.

**Comment by giordano** [comment]
I'm going to merge to let people test it more easily in https://github.com/JuliaVersionControl/Git.jl/pull/45, and this PR isn't any worse than the current situation anyway.

**Comment by maleadt** [comment]
I think this broke Git_jll for me on macOS: https://buildkite.com/julialang/metal-dot-jl/builds/251#01866c4e-d540-461b-af94-dca033cdc492

```
dyld[81854]: Library not loaded: @rpath/libpcre2-8.0.dylib
  Referenced from: /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/_git
  Reason: tried: '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/../lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/workspace/destdir/lib/libpcre2-8.0.dylib' (no such file), '/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/../lib/libpcre2-8.0.dylib' (no such file), '/libpcre2-8.0.dylib' (no such file)
ERROR: LoadError: failed process: Process(setenv(`/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/git -C /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/build/default-macmini-aarch64-5-0/julialang/metal-dot-jl/deps log -1 --format=%ct cmt`,["PATH=/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/41d2d4f312d93b56961767987cf48ea1f6d6fadc/bin:/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/b1a055a9c307b137556560efcc2cee3a6405dfd7/bin:/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin:/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/julia_installs/bin/mac/aarch64/1.8/julia-1.8-latest-macaarch64/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/Users/julia/.julia/artifacts/a81c923ea521c5513024eec57507055e4daa978a", "BUILDKITE_BUILD_CREATOR_TEAMS=", "BUILDKITE_LAST_HOOK_EXIT_STATUS=0", "BUILDKITE_PROJECT_PROVIDER=github", "GKS_ENCODING=utf8", "SECRET_CODECOV_TOKEN=[REDACTED]", "BASH_FUNC_nproc%%=() {  sysctl -n \"hw.perflevel0.logicalcpu\"\n}", "BUILDKITE_BUILD_CREATOR=github-actions[bot]", "BUILDKITE_REPO_MIRROR=/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/repos/https---github-com-JuliaGPU-Metal-jl-git", "BUILDKITE_AGENT_META_DATA_NUM_CPUS=4"  …  "BUILDKITE_AGENT_NAME=default-macmini-aarch64-5.0", "BUILDKITE_STEP_ID=01866c4e-d50b-40d3-afee-b485664c81a3", "CI=true", "BUILDKITE_STEP_KEY=", "BUILDKITE_TAG=", "BUILDKITE_LABEL=Julia 1.8", "BUILDKITE_COMMIT=bf5c3ca10d0fb172f4e586ff0b1b212c09920191", "BUILDKITE_PULL_REQUEST=99", "BUILDKITE_AGENT_PID=62594", "GIT_SSL_CAINFO=/Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-5.0/julia_installs/bin/mac/aarch64/1.8/julia-1.8-latest-macaarch64/share/julia/cert.pem"]), ProcessSignaled(6)) [0]
```

This is on the juliaecosystem CI machines, doing just `run('$(git()) ...')`.

**Comment by giordano** [comment]
You should really use Git.jl, this jll is quite useless anyway without that package or without setting all the environment variables that package sets anyway.

**Comment by maleadt** [comment]
I *am* using `Git.jl`, which depends on `Git_jll`.

**Comment by DilumAluthge** [comment]
Presumably we need to cut a new release of Git.jl that includes https://github.com/JuliaVersionControl/Git.jl/pull/45

**Comment by DilumAluthge** [comment]
https://github.com/JuliaRegistries/General/pull/78066

**Comment by maleadt** [comment]
With Git.jl 1.3.0 I'm getting  a different failure, again on macOS from `juliaecosystem`:

```
readlink: illegal option -- f
usage: readlink [-n] [file ...]
usage: dirname string [...]
/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/git: line 5: /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/metal-dot-jl/_git: No such file or directory
/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/git: line 5: exec: /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/metal-dot-jl/_git: cannot execute: No such file or directory
ERROR: LoadError: failed process: Process(setenv(`/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin/git -C /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/metal-dot-jl/deps log -1 --format=%ct cmt`,["PATH=/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/41d2d4f312d93b56961767987cf48ea1f6d6fadc/bin:/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/b1a055a9c307b137556560efcc2cee3a6405dfd7/bin:/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/depots/5cd495a2-4a16-4674-ae02-c839447744bb/artifacts/6a8cdbf9e758f66cee5f4f87fa20eb9be7521865/bin:/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/julia_installs/bin/macos/aarch64/1.9/julia-latest-macos-aarch64/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/Users/administrator/.julia/artifacts/a81c923ea521c5513024eec57507055e4daa978a", "BUILDKITE_BUILD_CREATOR_TEAMS=committers-at-julialang-slash-committers", "BUILDKITE_LAST_HOOK_EXIT_STATUS=0", "BUILDKITE_PROJECT_PROVIDER=github", "GKS_ENCODING=utf8", "SECRET_CODECOV_TOKEN=[REDACTED]", "BASH_FUNC_nproc%%=() {  sysctl -n \"hw.perflevel0.logicalcpu\"\n}", "BUILDKITE_BUILD_CREATOR=Tim Besard", "BUILDKITE_REPO_MIRROR=/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/repos/https---github-com-JuliaGPU-Metal-jl-git", "BUILDKITE_AGENT_META_DATA_NUM_CPUS=4"  …  "BUILDKITE_STEP_ID=01867a01-95ea-4f86-b65c-2e224efc44dc", "CI=true", "BUILDKITE_STEP_KEY=", "BUILDKITE_TAG=", "BUILDKITE_LABEL=Julia 1.9", "BUILDKITE_COMMIT=dd2be50b81e793c47b19592d4baee5c84ecaaf53", "BUILDKITE_PULL_REQUEST=85", "BUILDKITE_AGENT_PID=37308", "GIT_SSL_CAINFO=/Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/julia_installs/bin/macos/aarch64/1.9/julia-latest-macos-aarch64/share/julia/cert.pem", "OPENBLAS_MAIN_FREE=1"]), ProcessExited(126)) [126]
```

**Comment by giordano** [comment]
Right, `readlink -f` is probably GNU-only.  Do you know what's a good replacement for the BSD tools?

**Comment by giordano** [comment]
Actually, I'm not sure `-f` is really _needed_, it was more an extra goody.

**Comment by maleadt** [comment]
> Right, `readlink -f` is probably GNU-only. Do you know what's a good replacement for the BSD tools?

Sorry, no. Stack Overflow has some suggestions, https://stackoverflow.com/questions/1055671/how-can-i-get-the-behavior-of-gnus-readlink-f-on-a-mac, but yeah it'd be better if we can just drop it :-)

**Pull request merged**